allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -53,7 +80,16 @@ def __iter__(self) -> Iterator[Instance]:
                 raise ConfigurationError(
                     "For a lazy dataset reader, _read() must return a generator"
                 )
-            yield from instances
+            for instance in instances:
+                if self.vocab is not None:
+                    instance.index_fields(self.vocab)
+                yield instance
+
+    def index_with(self, vocab: Vocabulary):
+        self.vocab = vocab
+
+    def __len__(self):
+        return 1
```

Comment:
Why does it return 1?

Reply:
Yeah.... This is just moving previous behaviour from the iterator onto the dataset instead:

https://github.com/allenai/allennlp/blob/master/allennlp/data/iterators/data_iterator.py#L305

We rely in a couple of places that calling `len` on the iterator (or now, the dataloader) doesn't raise an error. In the case that you have an `IterableDataset` and you call `len`, the pytorch dataloader actually spits out a warning - but we need actually calling it to not crash.

Reply:
https://github.com/allenai/allennlp/blob/465224f867701f83ec3866b5913cfb96e74a4e80/allennlp/data/iterators/data_iterator.py#L314

I see. Maybe you can add a comment there?